ports/nrf/modules/machine/uart.c

```diff
@@ -321,29 +334,38 @@ STATIC const mp_rom_map_elem_t machine_hard_uart_locals_dict_table[] = {
 STATIC MP_DEFINE_CONST_DICT(machine_hard_uart_locals_dict, machine_hard_uart_locals_dict_table);
 
 STATIC mp_uint_t machine_hard_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t size, int *errcode) {
-    const machine_hard_uart_obj_t *self = self_in;
+    machine_hard_uart_obj_t *self = self_in;
     byte *buf = buf_in;
+    uint32_t t = self->timeout + mp_hal_ticks_ms();
 
     // read the data
     for (size_t i = 0; i < size; i++) {
         while (!uart_rx_any(self)) {
+            if ((mp_hal_ticks_ms() - t) >= self->timeout) {  // timed out
```

Comment:
I don't think this is correct: eg if no chars are available after `self->timeout` then the LHS here is 0 (or maybe 1) but the condition is still false.  So it should read `(int32_t)(mp_hal_ticks_ms() - t) >= 0`.  The cast to `int32_t` is very important.

Reply:
You're right.